### DISCUSSION TITLE
How to get client side routing using the new /app directory feature?

### DISCUSSION THREAD

**Discussion by igortrinidad** [vercel/next.js#50710] [App Router]
Hello there,

I'm playing arount the new App Router and app directory feature with layouts per page and i'm stucking on how to make route changes on frontend side only, without the need of the whole page load. 

What i want is something more related to a SPA page change than a PHP navigation...

This is some example just to show how the application using the navigation works:

[https://test-new-app-dir-structure.vercel.app](https://test-new-app-dir-structure.vercel.app)

Note: i'm using next/cookies in the app/layouts.tsx to load a cookie that stores the dark mode preference to avoid the application blink the theme color on the refresh of the page...

**Comment by icyJoseph** [comment]
That's kind of odd... the `next/link` should handle doing SPA transitions, as you describe. I'll inspect your project.

**Reply by icyJoseph** [reply]
Right, so yeah, your app has definitely something wrong. It probably has something to do with your cookie handling, without looking at code it is hard to tell, but something is definitely wrong in your app. Using `next/link` should just work.

What I see is that every route returns 404 when pre-fetched. Like from `/` the `login` route returns 404, and vice-versa. I think when that happens a "hard" navigation is done, to try and get the page/asset to respond again.

**Comment by devjiwonchoi** [comment]
Additional Reference supporting @icyJoseph's answer.

The `Link` component provides `client-side navigation` by default, so your case seems to be weird.

> Link is a React component that extends the HTML <a> element to provide [prefetching](https://nextjs.org/docs/app/building-your-application/routing/linking-and-navigating#prefetching) and client-side navigation between routes. It is the primary way to navigate between routes in Next.js.

You can find details at [Server-Centric Routing with Client-side Navigation](https://nextjs.org/docs/app/building-your-application/routing#server-centric-routing-with-client-side-navigation) and [Linking and Navigating](https://nextjs.org/docs/app/building-your-application/routing/linking-and-navigating).

**Comment by igortrinidad** [comment]
@icyJoseph @devjiwonchoi thanks for your attention, i think the error was caused for some old dependecy that was left behind while i tried to update the Next to the latest version in order to use the new App directory feature and Layouts...

I'm suspecting that could be something related to the next-i18next package that i used on the app, because it stopped to work after the creation of the repository with updated dependecies...

So, i created this repository if you guys want to check it out how i'm handling the cookies on the server side to update the dark mode:

[https://github.com/igortrinidad/next-new-app-router-error-test](https://github.com/igortrinidad/next-new-app-router-error-test)

And that's the application running this repository:

[https://next-new-app-router-error-test.vercel.app/](https://next-new-app-router-error-test.vercel.app/)

I'm trying to add the localization / internationalization functionality back to the app but it's giving 404 on the root route (/) [https://github.com/i18next/next-i18next](https://github.com/i18next/next-i18next)...

**Reply by devjiwonchoi** [reply]
Could you provide your active repo also? The repo you are hosting https://test-new-app-dir-structure.vercel.app/.

**Reply by igortrinidad** [reply]
For sure, the original repo is inside a company scope, but here it is: [https://github.com/igortrinidad/next-app-router-navigation-error](https://github.com/igortrinidad/next-app-router-navigation-error) and the vercel endpoint running this repo [https://next-app-router-navigation-error-test-as-it-was.vercel.app/en](https://next-app-router-navigation-error-test-as-it-was.vercel.app/en)...

While copying the repostiory i noticed that this repo is using Next version 13.06, maybe it could be what is causing the navigation error...
